kobo/apps/hook/tasks.py

```diff
@@ -46,4 +55,104 @@ def retry_all_task(hooklogs_ids):
+    # Use `.first()` instead of `.get()`, because task can be duplicated in admin section
+    failures_reports_period_task = PeriodicTask.objects.filter(enabled=True, task=beat_schedule.get("task"))\
+        .order_by("-last_run_at").first()
+
+    if failures_reports_period_task:
+        
+        last_run_at = failures_reports_period_task.last_run_at
+        queryset = HookLog.objects.filter(hook__email_notification=True, status=HOOK_LOG_FAILED)
+        if last_run_at:
+            queryset = queryset.filter(date_modified__gte=last_run_at)
+        queryset = queryset.order_by("hook__asset__name", "hook__uid", "-date_modified")
+
+        # PeriodicTask are updated every 3 minutes (default).
+        # It means, if this task interval is less than 3 minutes, some data can be duplicated in emails.
+        # Setting `beat-sync-every` to 1, makes PeriodicTask to be updated before running the task.
+        # So, we need to update it manually.
+        # see: http://docs.celeryproject.org/en/latest/userguide/configuration.html#beat-sync-every
+        PeriodicTask.objects.filter(task=beat_schedule.get("task")).update(last_run_at=timezone.now())
+
+        records = {}
+        max_length = 0
+
+        # Prepare data for templates.
+        # All logs will be grouped under their respective asset and user.
+        for record in queryset:
+            # if user doesn't exist in dict, add him
+            if record.hook.asset.owner.id not in records:
+                records[record.hook.asset.owner.id] = {
+                    "username": record.hook.asset.owner.username,
+                    # language is in implemented yet.
+                    # TODO add language to user table in registration process
+                    "language": getattr(record.hook.asset.owner, "language", "en"),
+                    "email": record.hook.asset.owner.email,
+                    "assets": {}
+                }
+
+            # if asset doesn't exist in user's asset dict, add it
+            if record.hook.asset.id not in records[record.hook.asset.owner.id]["assets"]:
+                max_length = 0
+                records[record.hook.asset.owner.id]["assets"][record.hook.asset.id] = {
+                    "name": record.hook.asset.name,
+                    "max_length": 0,
+                    "logs": []
+                }
+
+            # Add log to corresponding asset and user
+            records[record.hook.asset.owner.id]["assets"][record.hook.asset.id]["logs"].append({
+                "hook_name": record.hook.name,
+                "uid": record.uid,
+                "date_modified": record.date_modified,
+                "status_code": record.status_code,
+                "message": record.message
+            })
+            hook_name_length = len(record.hook.name)
+
+            # Max Length is used for plain text template. To display fixed size columns.
+            max_length = max_length if max_length > hook_name_length else hook_name_length
```

Comment:
could be `max_length = max(max_length, hook_name_length)`, right?

Reply:
Totally right!